library/core/src/num/dec2flt/common.rs

```diff
@@ -0,0 +1,195 @@
+    fn check_len(&self, n: usize) -> bool {
+        n <= self.as_ref().len()
+    }
+
+    /// Check if the first character in the slice is equal to c.
+    fn first_is(&self, c: u8) -> bool {
+        self.as_ref().first() == Some(&c)
+    }
+
+    /// Check if the first character in the slice is equal to c1 or c2.
+    fn first_is2(&self, c1: u8, c2: u8) -> bool {
+        if let Some(&c) = self.as_ref().first() { c == c1 || c == c2 } else { false }
+    }
+
+    /// Bounds-checked test if the first character in the slice is a digit.
+    fn first_isdigit(&self) -> bool {
+        if let Some(&c) = self.as_ref().first() { c.is_ascii_digit() } else { false }
+    }
+
+    /// Check if self starts with u with a case-insensitive comparison.
+    fn eq_ignore_case(&self, u: &[u8]) -> bool {
+        debug_assert!(self.as_ref().len() >= u.len());
+        let iter = self.as_ref().iter().zip(u.iter());
+        let d = iter.fold(0, |i, (&x, &y)| i | (x ^ y));
+        d == 0 || d == 32
+    }
+
+    /// Get the remaining slice after the first N elements.
+    fn advance(&self, n: usize) -> &[u8] {
+        &self.as_ref()[n..]
+    }
+
+    /// Get the slice after skipping all leading characters equal c.
+    fn skip_chars(&self, c: u8) -> &[u8] {
+        let mut s = self.as_ref();
+        while s.first_is(c) {
+            s = s.advance(1);
+        }
+        s
+    }
+
+    /// Get the slice after skipping all leading characters equal c1 or c2.
+    fn skip_chars2(&self, c1: u8, c2: u8) -> &[u8] {
+        let mut s = self.as_ref();
+        while s.first_is2(c1, c2) {
+            s = s.advance(1);
+        }
+        s
+    }
+
+    /// Read 8 bytes as a 64-bit integer in little-endian order.
+    unsafe fn read_u64_unchecked(&self) -> u64 {
+        debug_assert!(self.check_len(8));
+        let src = self.as_ref().as_ptr() as *const u64;
+        // SAFETY: safe as long as self is at least 8 bytes
+        u64::from_le(unsafe { ptr::read_unaligned(src) })
+    }
+
+    /// Try to read the next 8 bytes from the slice.
+    fn read_u64(&self) -> Option<u64> {
+        if self.check_len(8) {
+            // SAFETY: self must be at least 8 bytes.
+            Some(unsafe { self.read_u64_unchecked() })
+        } else {
+            None
+        }
+    }
+
+    /// Calculate the offset of slice from another.
+    fn offset_from(&self, other: &Self) -> isize {
+        other.as_ref().len() as isize - self.as_ref().len() as isize
+    }
+}
+
+impl ByteSlice for [u8] {}
+
+/// Helper methods to process mutable bytes.
+pub trait ByteSliceMut: AsMut<[u8]> {
+    /// Write a 64-bit integer as 8 bytes in little-endian order.
+    unsafe fn write_u64_unchecked(&mut self, value: u64) {
+        debug_assert!(self.as_mut().len() >= 8);
+        let dst = self.as_mut().as_mut_ptr() as *mut u64;
+        // SAFETY: safe as long as self is at least 8 bytes
+        unsafe {
+            ptr::write_unaligned(dst, u64::to_le(value));
+        }
```

Comment:
Is this true though? `self` is a reference to a slice of `u8`, which has a type-size of 1 and therefore an alignment of 1, and therefore a valid index may be at an odd address in memory. Meanwhile, `u64` has a type size of 8, and therefore often has an alignment of 8, which indeed is the cases on x86_64.

This should be a classic case of unaligned [access](https://www.kernel.org/doc/html/latest/core-api/unaligned-memory-access.html#avoiding-unaligned-accesses). The performance benefits here on any ARM architecture might be dubious, since ARM has major performance penalties for unaligned access, but this is **definitely** unaligned access.

Quick code for testing analogous C++ code is:

```cpp
// Compile with -Wcast-align=strict and -fsanitize=alignment

#include <cstdint>
#include <cstring>
#include <iostream>

int main() {
    std::cout << "alignof(u8)=" << alignof(uint8_t) << std::endl;
    std::cout << "alignof(u16)=" << alignof(uint16_t) << std::endl;
    std::cout << "alignof(u32)=" << alignof(uint32_t) << std::endl;
    std::cout << "alignof(u64)=" << alignof(uint64_t) << std::endl;

    uint8_t array[] = {
        52, 54, 53, 56, 55, 48, 54, 54, 56, 55, 54, 52, 53, 50, 54, 53, 49,
        48, 53, 49, 55, 48, 48, 48, 48, 51, 55, 50, 49, 55, 52, 52, 56, 48,
        50, 54, 51, 49, 50, 54, 51, 54, 53, 49, 53, 51, 53, 52, 52, 53, 56,
        51, 54, 56, 49, 53, 48, 51, 53, 54, 48, 53, 50, 53
    };
    uint8_t* ptr = array;
    uintptr_t address = (uintptr_t)ptr;
    if (address % 2 == 0) {
        address++;
        ptr++;
    }

    uint64_t value = *(uint64_t*)ptr;
    uint64_t expected;
    std::memcpy(&expected, ptr, 8);
    std::cout << "address=" << address << std::endl;
    std::cout << "value=" << value << ", expected=" << expected << std::endl;

    return 0;
}
```

Running this gives us the expect warning during compilation and failure at runtime:

```text
alignof(u8)=1
alignof(u16)=2
alignof(u32)=4
alignof(u64)=8
main.cc:36:14: runtime error: load of misaligned address 0x7ffd4dc0ce21 for type 'uint64_t', which requires 8 byte alignment
0x7ffd4dc0ce21: note: pointer points here
 7f 00 00  34 36 35 38 37 30 36 36  38 37 36 34 35 32 36 35  31 30 35 31 37 30 30 30  30 33 37 32 31
              ^ 
address=140725907934753
value=4050484495579100470, expected=4050484495579100470
```

Reply:
Also, just an FYI: thanks for spending so much time reviewing this. Your help has been immensely useful.

Reply:
`ptr::write_unaligned` has been changed to `ptr::write` in the latest [commit](https://github.com/Alexhuszagh/rust/blob/870ed2163282b2e5fa99585e6c28c916e4861646/library/core/src/num/dec2flt/common.rs#L101-L104).

Reply:
Actually, this is wrong. `write` can't be used since `dst` may not be [aligned](https://doc.rust-lang.org/std/ptr/fn.write.html). Miri correctly warns about alignment issues, which could cause crashes or major performance penalties for ARM architectures (and more). Reverting this back.